pkg/cosign/verify_blob.go

```diff
@@ -0,0 +1,223 @@
+	}, nil
+}
+
+func (s *blobSignature) Annotations() (map[string]string, error) {
+	return nil, nil
+}
+
+func (s *blobSignature) Payload() ([]byte, error) {
+	return s.payload, nil
+}
+
+func (s *blobSignature) Base64Signature() (string, error) {
+	return s.b64sig, nil
+}
+
+func (s *blobSignature) Cert() (*x509.Certificate, error) {
+	return nil, errors.New("no cert in blobSignature")
+}
+
+func (s *blobSignature) Chain() ([]*x509.Certificate, error) {
+	return nil, errors.New("no cert chain in blobSignature")
+}
+
+func (s *blobSignature) Bundle() (*bundle.RekorBundle, error) {
+	return s.bundle, nil
+}
+
+// VerifyBlobSignature verifies a signature
+func VerifyBlobSignature(ctx context.Context, blobBytes []byte, b64sig string, bundle *bundle.RekorBundle, co *CheckOpts) (checkedSignatures []oci.Signature, bundleVerified bool, err error) {
+	sig, err := newBundleSignature(blobBytes, b64sig, bundle)
+	if err != nil {
+		return checkedSignatures, bundleVerified, errors.Wrap(err, "failed to create bundle signature")
+	}
+
+	verifiers := []signature.Verifier{}
+	if co.SigVerifier != nil {
+		verifiers = append(verifiers, co.SigVerifier)
+	} else {
+		uuids, err := FindTLogEntriesByPayload(ctx, co.RekorClient, blobBytes)
+		if err != nil {
+			return checkedSignatures, bundleVerified, errors.Wrap(err, "failed to get tlog entries by payload")
+		}
+		if len(uuids) == 0 {
+			return checkedSignatures, bundleVerified, errors.New("could not find a tlog entry for provided blob")
+		}
+		for _, u := range uuids {
+			tlogEntry, err := GetTlogEntry(ctx, co.RekorClient, u)
+			if err != nil {
+				continue
+			}
+			certs, err := extractCerts(tlogEntry)
+			if err != nil {
+				continue
+			}
+			cert := certs[0]
+			verifier, err := ValidateAndUnpackCert(cert, co)
+			if err != nil {
+				continue
+			}
+			// Use the DSSE verifier if the payload is a DSSE with the In-Toto format.
+			if isIntotoDSSE(blobBytes) {
+				verifier = dsse.WrapVerifier(verifier)
+			}
+			verifiers = append(verifiers, verifier)
+		}
+	}
+
+	var validSigExists bool
+	for _, verifier := range verifiers {
+		if err := verifyOCISignature(ctx, verifier, sig); err != nil {
+			continue
+		}
+		bundleVerified = true
+		validSigExists = true
+		break
+	}
+	if !validSigExists {
+		fmt.Fprintln(os.Stderr, `WARNING: No valid entries were found in rekor to verify this blob.
+Transparency log support for blobs is experimental, and occasionally an entry isn't found even if one exists.
+We recommend requesting the certificate/signature from the original signer of this blob and manually verifying with cosign verify-blob --cert [cert] --signature [signature].`)
+		return checkedSignatures, bundleVerified, fmt.Errorf("could not find a valid tlog entry for provided blob, found %d invalid entries", len(verifiers))
+	}
+
+	if validSigExists {
+		fmt.Fprintln(os.Stderr, "Verified OK")
+		checkedSignatures = append(checkedSignatures, sig)
+	}
+
+	if !validSigExists && co.RekorClient != nil {
```

Comment:
How come the TLOG entry time will only be checked if a valid sig does not exist? `verifyOCISignature` does not check cert expiry time.

Reply:
Thank you for the information, I think we should have `VerifyBundle()` after `verifyOCISignature()` as VerifyImageSignature does. I will fix this.